File: ga/glm4_result_2024_6/result/multi_thread_basic_glm/7_multi_thread_basic.c
use std::sync::{Arc, Mutex};
use std::thread;
use std::cmp;

fn withdraw(account: Arc<Mutex<i32>>, amount: i32) {
    let mut balance = account.lock().unwrap();
    for _ in 0..amount {
        *balance = cmp::max(0, *balance - 1);
    }
}

fn main() {
    let account_balance = Arc::new(Mutex::new(1000000));
    let withdrawal_amount = 500000;

    let account_balance_clone = Arc::clone(&account_balance);
    let handle1 = thread::spawn(move || {
        withdraw(account_balance_clone, withdrawal_amount);
    });

    let account_balance_clone = Arc::clone(&account_balance);
    let handle2 = thread::spawn(move || {
        withdraw(account_balance_clone, withdrawal_amount);
    });

    handle1.join().unwrap();
    handle2.join().unwrap();

    println!("Final account balance: {}", *account_balance.lock().unwrap());
}

File: ga/gpt3_5_result_2024_6/result/multi_thread_basic_gpt/7_multi_thread_basic.rs
use std::thread;
use std::sync::{Mutex, Arc};

// Global shared variable representing account balance
static ACCOUNT_BALANCE: Mutex<i32> = Mutex::new(1000000);

fn withdraw() {
    let amount = 500000;
    for _ in 0..amount {
        let mut balance = ACCOUNT_BALANCE.lock().unwrap();
        *balance -= 1;
    }
}

fn main() {
    let balance = Arc::new(ACCOUNT_BALANCE);

    let balance_clone1 = Arc::clone(&balance);
    let handle1 = thread::spawn(move || {
        withdraw(balance_clone1);
    });

    let balance_clone2 = Arc::clone(&balance);
    let handle2 = thread::spawn(move || {
        withdraw(balance_clone2);
    });

    handle1.join().unwrap();
    handle2.join().unwrap();

    println!("Final account balance: {}", *balance.lock().unwrap());
}